### ISSUE TITLE
Value does not fall within the expected range

### ISSUE THREAD

**Issue by SimonCropp** [jbevain/cecil#685]
@jbevain still having fun repro'ing this in a way i can properly debug. i will add more info to this issue as i find it

Stacktrace:
```
Mono.Cecil.Pdb.SymWriter.OpenScope(int startOffset) Line 146	C#
Mono.Cecil.Pdb.NativePdbWriter.DefineScope(Mono.Cecil.Cil.ScopeDebugInformation scope, Mono.Cecil.Cil.MethodDebugInformation info, out Mono.Cecil.MetadataToken import_parent) Line 149	C#
Mono.Cecil.Pdb.NativePdbWriter.DefineScope(Mono.Cecil.Cil.ScopeDebugInformation scope, Mono.Cecil.Cil.MethodDebugInformation info, out Mono.Cecil.MetadataToken import_parent) Line 191	C#
Mono.Cecil.Pdb.NativePdbWriter.Write(Mono.Cecil.Cil.MethodDebugInformation info) Line 69	C#
Mono.Cecil.Cil.CodeWriter.WriteResolvedMethodBody(Mono.Cecil.MethodDefinition method) Line 134	C#
Mono.Cecil.Cil.CodeWriter.WriteMethodBody(Mono.Cecil.MethodDefinition method) Line 54	C#
Mono.Cecil.MetadataBuilder.AddMethod(Mono.Cecil.MethodDefinition method) Line 1658	C#
Mono.Cecil.MetadataBuilder.AddMethods(Mono.Cecil.TypeDefinition type) Line 1651	C#
Mono.Cecil.MetadataBuilder.AddType(Mono.Cecil.TypeDefinition type) Line 1440	C#
Mono.Cecil.MetadataBuilder.AddTypes() Line 1413	C#
Mono.Cecil.MetadataBuilder.BuildTypes() Line 1266	C#
Mono.Cecil.MetadataBuilder.BuildModule() Line 1036	C#
Mono.Cecil.MetadataBuilder.BuildMetadata() Line 1006	C#
Mono.Cecil.ModuleWriter.BuildMetadata.AnonymousMethod__2_0(Mono.Cecil.MetadataBuilder builder, Mono.Cecil.MetadataReader _) Line 144	C#
Mono.Cecil.ModuleDefinition.Read<Mono.Cecil.MetadataBuilder, Mono.Cecil.MetadataBuilder>(Mono.Cecil.MetadataBuilder item, System.Func<Mono.Cecil.MetadataBuilder, Mono.Cecil.MetadataReader, Mono.Cecil.MetadataBuilder> read) Line 955	C#
Mono.Cecil.ModuleWriter.BuildMetadata(Mono.Cecil.ModuleDefinition module, Mono.Cecil.MetadataBuilder metadata) Line 143	C#
Mono.Cecil.ModuleWriter.Write(Mono.Cecil.ModuleDefinition module, Mono.Disposable<System.IO.Stream> stream, Mono.Cecil.WriterParameters parameters) Line 119	C#
Mono.Cecil.ModuleWriter.WriteModule(Mono.Cecil.ModuleDefinition module, Mono.Disposable<System.IO.Stream> stream, Mono.Cecil.WriterParameters parameters) Line 78	C#
Mono.Cecil.ModuleDefinition.Write(string fileName, Mono.Cecil.WriterParameters parameters) Line 1144	C#	
```

offending variable
![image](https://user-images.githubusercontent.com/122666/89722469-0fb31700-da2d-11ea-98df-4b7ae19c0ceb.png)

**Comment by SimonCropp** [comment]
ok if i remove these line https://github.com/jbevain/cecil/pull/677/files#diff-37bd441e63179a2b67548a97a281c3adR256-R261 from  InstructionCollection.OnRemove it starts working again 


```
protected override void OnRemove (Instruction item, int index)
{
	var previous = item.previous;
	if (previous != null)
		previous.next = item.next;

	var next = item.next;
	if (next != null)
		next.previous = item.previous;

	RemoveSequencePoint (item);

-	var scope = GetLocalScope ();
-		if (scope != null) {
-			var size = item.GetSize ();
-			UpdateLocalScope (scope, item.Offset + size, -size, item);
-		}

	item.previous = null;
	item.next = null;
}
```

any ideas?

**Comment by SimonCropp** [comment]
i also ran a smoke test on ~30 fody weavers with the InstructionCollection.OnRemove tweak and they worked.

**Comment by SimonCropp** [comment]
it also only happens when the target assembly is in release mode

**Comment by SimonCropp** [comment]
also only happens with DebugType = full or pdbonly.  embedded, portable and none all work.

**Comment by jbevain** [comment]
@vitek-karas any insight here?

**Comment by vitek-karas** [comment]
@SimonCropp sorry I'm a bit new to Cecil repo - how would I repro this? Take the code example and somehow run it through read/write loop in Cecil? Or something else? Given the bug I doubt that will work - I assume it needs some IL manipulation.

**Comment by SimonCropp** [comment]
i havnt been able to get an isolated repro. do you want to jump on a call and i can show you?

**Comment by vitek-karas** [comment]
Sure - where/how

**Comment by SimonCropp** [comment]
what timezone are you in?

**Comment by vitek-karas** [comment]
Europe - if you have time now... it would work great for me.

**Comment by vitek-karas** [comment]
I was looking into the code some more to try to guess what could be wrong. In general I'm pretty sure that before my change the resulting PDB would be wrong anyway, it just might not crash (by chance). We actually found the problem by running the resulting PDB through a PDB processing tool (convert from native PDB to managed PDB) and it crashed the native PDB reader.

I made some improvements to try to make sure that scopes are valid (not correct all the time, but at least more valid): https://github.com/vitek-karas/cecil/tree/GuaranteeValidScopes

Can you please give it a try? I still can't come up with a scenario how to get what I think is your behavior, but I could come up with a scenario where the scopes would be invalid (and I assume the writer would fail on them).

Some more on the problem of local scopes when the method body is editted:
* Cecil stores the local scopes in two possible ways
  * Start and End are offsets (in bytes from start of method body) - this is how scopes are loaded from disk for all methods
  * Start and End are pointers to the Instruction objects (this is how a scope would be created in code)
* Each instruction stores its offset - this is populated when loaded from disk
* As method body is editted, offsets on instructions are not updated (and offsets on scopes were not updated either before my change)
* Newly created instructions get offset 0
* Before writing the assembly all instruction offsets are recalculated

So this means that before my change on write instructions had correct offsets, but the scopes were saved as loaded and so removing instructions could easily create invalid scopes. Also really any edit of the method made the scopes wrong.
After my change if instructions are only removed everything should be "correct". If new instructions are added things are still broken - just somewhat differently I guess.

My change works basically only on scopes which store offsets (so those loaded from disk), scopes which are created manually should not be affected. But in general my changes either keeps scopes stored by offset, or converts pointers to offsets when absolutely necessary.

There's an alternative way to "Fix" this - when editing method body, convert all scopes to pointers (no offsets). This would require full walk of all instructions. After that edits get easy (no more offset math, just pointer moves) and the scopes are basically guaranteed to be correct. I didn't want to go with this change because it's rather big - if somebody has code which uses the offsets on scopes it would like change behavior - even for scopes which are not modified at all. So I went with the offset approach which tries to only affect scopes which were affected by edits.

**Comment by vitek-karas** [comment]
I implemented the more robust version of the local scopes update which force-resolves all local scope offsets before updating them according to the method body edit. I also added better tests to validate that PDB reader/writer can handle the results.

As far as I can tell, loading the assembly and symbols with PDB reader will pretty much always return resolved local scopes, the change should have minimal impact. It's biggest impact is in cases where the local scopes are not resolved, in which case it resolves them - which will make the `Instruction` return the correct value (not-null), but it may return wrong `Offset` since the offset now comes from the instruction which can be wrong (after edits). Before saving to disk all instruction offsets are updated, so upon save the values will be correct, but if something relies on the offset values the way they behave currently, this may break.

https://github.com/jbevain/cecil/pull/687

It would be great if you could try the new fix against your repro to see if it resolves the problem.

**Comment by jbevain** [comment]
Fixed by #687
